README.md

```diff
@@ -68,6 +93,8 @@ Create a new async iterable. The values yielded from calls to `.next()` or when
 `options` is an _optional_ parameter, an object with the following properties:
 
 * `onEnd` - a function called after _all_ values have been yielded from the iterator (including buffered values). In the case when the iterator is ended with an error it will be passed the error as a parameter.
+* `objectMode` - a boolean value that means non-`Uint8Array`s will be passed to `.push`, default: `false`
+* `highWaterMark` - a threshold beyond which calls to `.push` will wait for previously pushed items to be consumed.  If `objectMode` is `false`, this number is interpreted as the total number of bytes that the queue should hold, otherwise it is the number of items in the queue
```

Comment:
> The `.push(value)` and `.end([err])` docs above need an update to specify that they return a promise and what it means when you await on that promise.
>
> Awaiting on a `push` does _not_ mean it has been consumed (which is what I think I would expect) so I think needs to be clarified!

Will do.

> What happens if you don't await on the `push` - is it effectively ignoring the highwater mark (buffering everything)?

Internally it'll still wait for space in the queue, but the promise will go onto the heap and I guess may explode if you have a very large number of items queue as they will eat memory (this has a similar effect to the current release version).  Your linter should also scream at you for having [floating promises](https://www.npmjs.com/package/eslint-plugin-no-floating-promise) 😆 .

> I'm not sure I understand why `end` returns a promise - under what curcumstances would I await on it?

It is a bit weird, you are right - maybe it'd be better if it just synchronously closed the pushable for writing so any subsequent calls to `.push` would throw?

If an error is passed it could be cached until the queue empties (so values in pushes that are waiting for queue space can still be consumed) then thrown on the next call to `.next`, or do you think it should reject all pending `.push` promises and throw immediately?

I get the feeling it should fail fast but I'm not sure.

> If nothing ever drains the source once the highwater mark is reached, what happens?

Subsequent calls to `.push` will wait until there's space in the queue before resolving so your application will hang in the same way as if you forget to add a listener for the `'data'` event of a regular node stream.

I guess we need a way to read if the pushable has space in it's queue, similar to `writableLength` in node streams.

> Do we need a way to cancel a push? 

Probably - an optional `AbortSignal` argument could be added, that should do it?

Reply:
> or do you think it should reject all pending .push promises and throw immediately?

Having said that this is in the readme:

>  If an error is passed the buffer is cleared immediately and the next iteration will throw the passed error

So I guess yeah, it should reject any pending promises and throw the passed error.